src/brogue/IO.c

```diff
@@ -3050,24 +3050,90 @@ int formatCountedMessage(char *buffer, size_t size, archivedMessage *m) {
     return length;
 }
 
-// Fill the given buffer, which is height lines tall and COL*2 characters wide, with archived messages.
-// Fill from the bottom, so that the most recent message appears in the last line of buf.
-void formatRecentMessages(char buf[][COLS*2], size_t height) {
-    size_t i;
+// Copy \n-delimited lines to the given buffer.  Carry colors across line breaks.
+void splitLines(short lines, char wrapped[COLS*20], char buffer[][COLS*2], short bufferCursor) {
+    short linesSeen;
+    char color[5], line[COLS*2];
+    char *start, *end;
+
+    start = &(wrapped[0]);
+    color[0] = '\0';
+    line[0] = '\0';
+
+    for (end = start, linesSeen = 0; *end; end++) {
+        if (*end == COLOR_ESCAPE) {
+            strncpy(color, end, 4);
+            color[4] = '\0';
+            end += 4;
+        } else if (*end == '\n') {
+            *end = '\0';
+
+            if (bufferCursor + 1 - lines + linesSeen >= 0) {
+                strncpy(line, color, 5);
+                strncat(line, start, COLS*2 - strlen(line) - 1);
+                line[COLS*2-1] = '\0';
+                strncpy(buffer[bufferCursor + 1 - lines + linesSeen], line, COLS*2);
+                line[0] = '\0';
+            }
 
-    for (i = 0; i < height; i++) {
-        formatCountedMessage(buf[i], COLS*2, getArchivedMessage(height - i));
+            linesSeen++;
+            start = end + 1;
+        }
+    }
+
+    strncpy(line, color, 5);
+    strncat(line, start, COLS*2 - strlen(line) - 1);
+    line[COLS*2-1] = '\0';
+    strncpy(buffer[bufferCursor], line, COLS*2);
+}
+
+// Fill the buffer of height lines with archived messages.  Fill from the
+// bottom, so that the most recent message appears in the last line of buffer.
+void formatRecentMessages(char buffer[][COLS*2], size_t height, short *linesFormatted, short *latestMessageLines) {
+    short lines, bufferCursor, messagesFormatted;
+    char folded[COLS*2], wrapped[COLS*2];
+    archivedMessage *nextMessage;
+
+    bufferCursor = height - 1;
+    messagesFormatted = 0;
+
+    while (bufferCursor >= 0 && messagesFormatted < MESSAGE_ARCHIVE_ENTRIES) {
+        nextMessage = getArchivedMessage(messagesFormatted + 1);
+        if (!nextMessage->message[0]) {
+            break;
+        }
+
+        formatCountedMessage(folded, COLS*2, nextMessage);
+
+        lines = wrapText(wrapped, folded, DCOLS);
+        splitLines(lines, wrapped, buffer, bufferCursor);
+
+        if (bufferCursor == height - 1 && latestMessageLines) {
+            *latestMessageLines = lines;
+        }
+
+        bufferCursor -= lines;
+        messagesFormatted++;
+    }
+
+    if (linesFormatted) {
+        *linesFormatted = height - 1 - bufferCursor;
+    }
+
+    while (bufferCursor >= 0) {
+        buffer[bufferCursor--][0] = '\0';
     }
 }
 
 // Display recent archived messages
-void displayRecentMessages() {
-    short i;
+void displayRecentMessages(boolean confirmLatest) {
```

Comment:
It's a pre-existing concept which I've moved around a bit.  Notice that in the normal three-line message area (when the `M` display is closed) that recently arrived messages are bold and bright, and a little later they are dimmed.  Such bright messages are "unconfirmed".

This `displayRecentMessages` function is now responsible for redrawing the three-line message area, and so wants to know if the "most recent" (whatever that means) messages should be bright ("are unconfirmed") or not.  I opted to expose it as a parameter.

Reply:
This feels like another function, some "confirmMessages", but i see the difficulty, because you can't know latest messages before you draw them

Reply:
..you know, there is a `confirmMessages`, which amounts to a call to `displayRecentMessages(true)`.

The original issue which led to `displayRecentMessages` was an inability to express `displayRecentMessages(false)` (`confirmMessages` can't do that.)

I may go and either remove `confirmMessages` or make it a delegate to `displayRecentMessages`.

Reply:
It almost feels like confirmation should be a part of the archedMessage type. Instead of being stored by line in a different global. This would untangle the interplay between displaying and confirming, also removing the need for latestMessageLines I think?

Reply:
The confirmation flags could certainly be moved into the archive itself (I suppose I would put it in `flags` as either a new `CONFIRMED` or perhaps `UNCONFIRMED` value.)

Currently this branch redefines "confirmed" to be `archivedMessage.turn < rogue.playerTurnNumber` instead of letting the `messageConfirmed` global be the source of truth.  So in some sense a `UNCONFIRMED` flag is redundant with `.turn` itself.

The `confirmMessages` function (which has a few sprinkled call sites) would then end up walking the archive, ensuring the flags are correct on each message back into history.  For a state that's only supposed to exist for a brief while before being clear essentially forever, that walking might feel a bit schlemiel-ish.

What we might do is just retire the `messageConfirmed` global completely.  I believe the branch is pretty close to being able to do that.  I'm all for taking globals out behind the chemical shed and "retiring" them, I just didn't want to bake in even more complexity off the bat.

Reply:
..I just realized my train of thought had an important flaw.  "Confirmed" is still a single bit that twiddles back and forth every turn as the player takes actions within a turn that dims the message display.

We couldn't retire `messageConfirmed`, but we could change it from an array back to a single flag.

Reply:
..ok.  And if `messageConfirmed` was a single global boolean, `displayRecentMessages` might reference it directly instead of accepting a parameter.

Then `confirmMessages`, which would continue to exist, would set the global flag to the appropriate state and then delegate to `displayRecentMessages`.

Reply:
> Currently this branch redefines "confirmed" to be archivedMessage.turn < rogue.playerTurnNumber instead of letting the messageConfirmed global be the source of truth

I hadn't fully realised this - in that case the argument to dRM makes sense: it's whether to ignore the lack of confirmation or not for just the drawn turn. In that case, yes I think it would be great to try and get this down to one concept, instead of the control flow (and where confirmLatest is true or false) having an implicit dependence on the turn progression in the game.

Reply:
Ok, I'll take another pass at cleaning it up.  I think I'll try the design I alluded to above (the array becomes a single boolean, cM still exists and defers to dRM, dRM loses its argument and just looks at the global boolean.)

Reply:
Well, instead of a boolean it became a short (`messagesUnconfirmed`), but otherwise that played out more or less as envisioned.